### ISSUE TITLE
XRPresentationContext attachment lifetime

### ISSUE THREAD

**Issue by toji** [immersive-web/webxr#375]
In the spec currently we specify that passing an `XRPresentationContext` in to `requestSession` establishes it's canvas as the in-page output surface for anything that's rendered with the session, which may be magic window content or HMD mirroring depending on context (and will soon serve as the output surface for Phone AR cases as well.) As a refresher the code looks like this:

```js
let xrCanvas = document.createElement('canvas');
let xrOutputContext = xrCanvas.getContext('xrpresent');
document.body.appendChild(xrCanvas);

// Request a non-immersive session for magic window rendering.
xrDevice.requestSession({ outputContext: xrOutputContext })
    .then(OnSessionStarted);
```

So far this pattern appears to be working well (some concerns about polyfill performance aside.) What we don't yet specify, though, is what should happen when the context is passed to multiple sessions. For example, it would be a convenience for developers if they could do something like so:

```js
let xrCanvas = document.createElement('canvas');
let xrOutputContext = xrCanvas.getContext('xrpresent');
document.body.appendChild(xrCanvas);

// Request a non-immersive session for magic window rendering.
xrDevice.requestSession({ outputContext: xrOutputContext })
    .then(OnSessionStarted);

function onXRButtonClicked() {
  // Request an immersive session with a mirror.
  xrDevice.requestSession({ immersive:true, outputContext: xrOutputContext })
    .then(OnSessionStarted);
}
```

It's common to want to show the same content in magic window and immersive modes, and as a result it's logical to want the page element to simply switch over from showing the magic window to showing the mirror and back again when appropriate. But that creates an ambiguity about how the output context is associated with the sessions. Does it stay attached to the session for life? Can it be "stolen" by a new session? Is attaching it to an immersive session a special case? We should decide and document it clearly.

I see a few options, and would be glad to hear more if you have them:

  1. `XRPresentationContext`s can only be bound to one `XRSession` for their lifetime. To produce the above described behavior you would create a new canvas/context pair for the immersive session and swap it out on the page on session start/end. More work for the developer and more opportunities to goof it up, but the behavior is extremely straightforward.
 2. `XRPresentationContext`s can only be bound to one `XRSession` at a time, but are transferred when passed to a new session. This is straightforward and easy to explain, at least, but it doesn't seem to have much benefit and several significant downsides. For example, unless we wanted to add a new method to late-bind output contexts (which I'd like to avoid) there would be no way to switch back to the original non-immersive session in the above scenario. In fact we'd probably have to end the session when the context was transferred. Effectively we'd be artificially limiting the lifetime of an expensive object (sessions) in order to prevent repeated creation of cheap objects (canvas/context). Not recommended.
 3. `XRPresentationContext`s can only be bound to one non-immersive `XRSession` at a time, but an immersive `XRSession` can "borrow" a context while active (since the non-immersive ones will be suspended anyway). This option is kind of janky spec-wise, but caters to making what is likely to be the most common use case easy.

For what it's worth, I think option 1 is the most practical, but I'm curious to hear what the other CG members think?

**Comment by toji** [comment]
Oh, there is a fourth option I can think of:

  4. `XRPresentationContext`s can only be bound to any number of `XRSession`s, and whichever session pushes a frame last "wins". (For the record, this happens to be what Chrome does today, but it was more of a naturally emerging behavior than anything intentional.) This gives us the magic window->immersive->magic window flow naturally and cleanly, because the magic window session will be suspended while the immersive session is running, and therefore only one is producing frames at a time. But if you passed it to two non-immersive sessions and ran their rAF loops simultaneously you'd either never see the output of one of the sessions or you'd get flickering between the two session's outputs.

**Comment by RafaelCintron** [comment]
One drawback with option 4 is that web developers may accidentally author content that works on some headset configurations but not others.  For instance, if the second session that attaches to the presentation context draws at a higher framerate than the first one, then the presentation context will always show the second session's rendering.  The developer may think their content has no bugs and will be surprised when their content flickers in the wild.  

If people feel the first option is too limiting, we can consider a variation of option 2 where the presentation context keeps a stack of connecting sessions.  If A connects, then B, then C, the presentation context will always display C.  When C ends, it will go back to displaying B.  If there is no use case for keep a stack, then I think the first option is the best, simplest approach.

**Comment by toji** [comment]
We talked about this on the WebXR call today. I generally don't mind Rafael's stack idea, even though it could lead to slightly magical-feeling behavior in some complex cases. There was some confusion in trying to explain this verbally, though, so lets walk through a quick example:

In this scenario each `XRPresentationContext` would have a stack of `XRSessions` that they're bound to, only accepting imagery from the top-most session. (Small clarification: it would have to be the top-most non-blurred session for this to be effective)

So if you start with

```
let xrCanvas = document.createElement('canvas');
let xrOutputContext = xrCanvas.getContext('xrpresent');
document.body.appendChild(xrCanvas);

// Request a non-immersive session for magic window rendering.
xrDevice.requestSession({ outputContext: xrOutputContext })
    .then((session) => {
  xrSession0 = session;
  // ...
});
```

`xrOutputContext` now has a stack of sessions that looks like:

```
[xrSession0]
```

If we then create an immersive session with the same context

```
function onXRButtonClicked() {
  // Request an immersive session with a mirror.
  xrDevice.requestSession({ immersive:true, outputContext: xrOutputContext })
    .then((session) => {
      xrSession1 = session;
      // ...
    });
}
```

The stack now looks like this:

```
[xrSession1,
 xrSession0 (blurred)]
```

And if that session is ended `xrSession1` gets popped off the top and `xrSession0` resumes presentation and using the context. This gives us the nice, smooth upgrade and downgrade experience that Nell advocated for in the call.

Where it gets complicated is if, instead of ending the immersive session you instead try to create another non-immersive session with the same context, at which point the stack would look like this:

```
[xrSession2, (blurred)
 xrSession1,
 xrSession0 (blurred)]
```

And ending the exclusive session would return control of the context to a session which never had it prior. Similarly, attempting to bind multiple non-exclusive sessions to a single context would create a "last one wins" behavior that users might find surprising.

To capture one other point made on the call: It's tempting to solve this issue with a `setOutputContext()` method on the session that has transfer semantics, but that allows us to have active non-immersive sessions which don't have an associated `outputContext`, which the group has previously determined was bad.

**Comment by blairmacintyre** [comment]
I admit that as I look through the code and read this, it leaves me feeling uneasy;  this seems very over-engineered, given the simple alternatives to the problem (as I understand it) we're trying to solve.

Developers are very likely going to have to do some changes to their internal state when we switch from magic window (or "just mouse-controlled 3d") to immersive mode and back.  These will be isolated in the frameworks, or done once.  Having to also recreate a session or re-set a rendering context seems trivial and similar to changing the control flow or event handlers, resetting the 3D camera, and so forth needed to deal with the flip back and forth.

So I wouldn't even be against something even simpler, where the non-immersive session becomes invalid when the immersive one is created, and the programmer recreates it when the user exits immersive mode.  Alternatively, a session could be suspended (i.e., blurred) and have a `reActivate` method that could be called to attempt to restart the session when the immersive mode is exited.

**Comment by toji** [comment]
I'm currently working on a proposal that may resolve this issue by allowing sessions to transition between inline and immersive modes, which then allows for the same outputContext to be used for both modes. That model seems vastly preferable to any of the above solutions which, I agree with @blairmacintyre, all seem over-engineered.

**Comment by judax** [comment]
I am in favor of using the same context if possible. The rationale is that for complex scenes, the rebuilding of shaders, the upload of textures, geometry, ... can be pretty cumbersome and lead to stalls in the app. So, eager to listen to your proposal @toji !

**Comment by toji** [comment]
Fixed by #504
